### PR TITLE
[MCCAS] Introduce zeroing-out of the abbreviation offsets

### PR DESCRIPTION
**Pull request by felipepiovezan**
This PR implements the required refactors as well as the zeroing out of abbreviation offsets inside the CUs of `__debug_abrev` sections

### CODE REVIEW

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h [hunk 1]
```diff
@@ -358,7 +358,7 @@ class MCCASBuilder {
   Error finalizeGroup();
 
   void startSection(const MCSection *Sec);
-  Error finalizeSection();
+  template <typename SectionRefTy = SectionRef> Error finalizeSection();
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.h:361]: @cachemeifyoucan this commit contains all the changes I needed (for now?) to add a new CAS object type for the debug info section

Comment (cachemeifyoucan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h:361]: The new section type looks fine to me. Can't understand how you extract abbr field by glimpsing through, some comments or explanation in commit message is appreciated.

Also you can run the benchmark to see the gain in llvm-project builds.

Comment (felipepiovezan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h:361]: >  Can't understand how you extract abbr field by glimpsing through, some comments or explanation in commit message is appreciated.

The core functionality for extracting abbreviation offsets was done in previous PRs, inside the function  `getAndSetDebugAbbrevOffsetAndSkip`.

The first commit slightly changes it to overwrite offsets. I'll update that particular message to emphasize the point that we are not changing _how_ the offset is found

**Review by adrian-prantl — APPROVED** [review]

**Review by rastogishubham — APPROVED** [review]
I think this PR looks good!

**Comment by felipepiovezan** [comment]
Addressed review comments, rebased.
@swift-ci  please test

**Comment by felipepiovezan** [comment]
@swift-ci please test

**Comment by felipepiovezan** [comment]
Added explicit move to `return E`. Not sure why this is required with C++17 too :/

**Comment by felipepiovezan** [comment]
@swift-ci  please test

**Comment by rastogishubham** [comment]
@swift-ci please test

**Pull request merged**